rskj-core/src/main/java/co/rsk/net/AsyncNodeBlockProcessor.java

```diff
@@ -0,0 +1,152 @@
+ * If a block passes validation, it will immediately be propagated to other nodes and its execution will be scheduled on a separate thread.
+ * Blocks are being executed and connected to the blockchain sequentially one after another.
+ * <p>
+ * If a block is not ready to be added to the blockchain, it will be on hold in a BlockStore.
+ * <p>
+ */
+public class AsyncNodeBlockProcessor extends NodeBlockProcessor implements InternalService, Runnable {
+
+    private static final Logger logger = LoggerFactory.getLogger("asyncblockprocessor");
+
+    private final BlockingQueue<PeerBlockPair> blocksToProcess = new LinkedBlockingQueue<>();
+
+    private final Thread thread = new Thread(this,"async block processor");
+
+    private final Listener listener;
+
+    private volatile boolean stopped;
+
+    public AsyncNodeBlockProcessor(@Nonnull NetBlockStore store, @Nonnull Blockchain blockchain, @Nonnull BlockNodeInformation nodeInformation,
+                                   @Nonnull BlockSyncService blockSyncService, @Nonnull SyncConfiguration syncConfiguration,
+                                   @Nullable Listener listener) {
+        super(store, blockchain, nodeInformation, blockSyncService, syncConfiguration);
+        this.listener = listener;
+    }
+
+    @Override
+    public BlockProcessResult processBlock(@Nullable Peer sender, @Nonnull Block block) {
+        // TODO: add block validation
+
+        BlockProcessResult blockProcessResult = blockSyncService.processBlock(block, sender, false, false);
+        if (blockProcessResult.isScheduledForProcessing()) {
+            blocksToProcess.offer(new PeerBlockPair(sender, block));
+        }
+
+        return blockProcessResult;
+    }
+
+    @Override
+    public void start() {
+        thread.start();
+    }
+
+    @Override
+    public void stop() {
+        stop(0L);
+    }
+
+    /**
+     * Stop the service and wait until a working thread is stopped for {@code waitMillis} milliseconds,
+     * if {@code waitMillis} greater than zero. If {@code waitMillis} is zero, then immediately returns.
+     */
+    public void stop(long waitMillis) {
+        stopped = true;
+
+        try {
+            thread.interrupt();
+            if (waitMillis > 0L) {
+                thread.join(waitMillis);
+            }
+        } catch (InterruptedException e) {
+            logger.error("Failed to join the thread", e);
+        }
+    }
+
+    @Override
+    public void run() {
+        while (!stopped) {
+            Peer sender = null;
+            Block block = null;
+
+            try {
+                logger.trace("Get peer/block pair");
+
+                PeerBlockPair pair = blocksToProcess.take();
+
+                sender = pair.peer;
+                block = pair.block;
+
+                logger.trace("Start block processing");
+                BlockProcessResult blockProcessResult = blockSyncService.processBlock(block, sender, false, true);
```

Comment:
Processing the block, connecting to a chain, OK. Any second relay of this block? AFAIR geth implementation sends a second relay after the full process of the block

Reply:
yes, as agreed this could be part of the next phase.